lib/lightning/version_control/version_control.ex

```diff
@@ -123,6 +127,197 @@ defmodule Lightning.VersionControl do
     end
   end
 
+  @doc """
+  Fetches the oauth access token using the code received from the callback url
+  For more info: https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-user-access-token-for-a-github-app
+  """
+  @spec exchange_code_for_oauth_token(code :: String.t()) ::
+          {:ok, map()} | {:error, map()}
+  def exchange_code_for_oauth_token(code) do
+    app_config = Application.fetch_env!(:lightning, :github_app)
+
+    query_params = [
+      client_id: app_config[:client_id],
+      client_secret: app_config[:client_secret],
+      code: code
+    ]
+
+    client = GithubClient.build_oauth_client()
+
+    case Tesla.post(client, "/access_token", %{}, query: query_params) do
+      {:ok, %{body: %{"access_token" => _} = body}} ->
+        {:ok, body}
+
+      {:ok, %{body: body}} ->
+        {:error, body}
+
+      other ->
+        other
+    end
+  end
+
+  @doc """
+  Fetches a new access token using the given refresh token
+  For more info: https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/refreshing-user-access-tokens
+  """
+  @spec refresh_oauth_token(refresh_token :: String.t()) ::
+          {:ok, map()} | {:error, map()}
+  def refresh_oauth_token(refresh_token) do
+    app_config = Application.fetch_env!(:lightning, :github_app)
+
+    query_params = [
+      client_id: app_config[:client_id],
+      client_secret: app_config[:client_secret],
+      grant_type: "refresh_token",
+      refresh_token: refresh_token
+    ]
+
+    client = GithubClient.build_oauth_client()
+
+    case Tesla.post(client, "/access_token", %{}, query: query_params) do
+      {:ok, %{body: %{"access_token" => _} = body}} ->
+        {:ok, body}
+
+      {:ok, %{body: body}} ->
+        {:error, body}
+
+      other ->
+        other
+    end
+  end
+
+  @doc """
+  Checks if the given token has expired.
+  Github supports access tokens that expire and those that don't.
+  If the `access token` expires, then a `refresh token` is also availed.
+  This function simply checks if the token has a `refresh_token`, if yes, it proceeds to check the expiry date
+  """
+  @spec oauth_token_valid?(token :: map()) :: boolean()
+  def oauth_token_valid?(token) do
+    case token do
+      %{"refresh_token_expires_at" => expiry} ->
+        {:ok, expiry, _offset} = DateTime.from_iso8601(expiry)
+        now = DateTime.utc_now()
+        DateTime.after?(expiry, now)
+
+      %{"access_token" => _access_token} ->
+        true
+
+      _other ->
+        false
+    end
+  end
+
+  @doc """
+  Fecthes the access token for the given `User`.
+  If the access token has expired, it `refreshes` the token and updates the `User` column accordingly
+  """
+  @spec fetch_user_access_token(User.t()) ::
+          {:ok, String.t()} | {:error, :expired | map()}
```

Comment:
The spec is a bit different `{:error, :expired}`. Could you double check the error handling?